[BE] Team-12: To-Do List #1

Malloc72P opened the pull request:
안녕하세요! 12팀의 BE를 맡은 NAS와 Jung이라고 합니다!

이번에는 팀 활동에 중점을 두고 작업했어요!

작업 내용을 요약하면 아래와 같습니다!

## 1. 구현 상황
* API설계에 앞서서 용어정리를 했습니다! [위키 링크](https://github.com/Malloc72P/todo-list/wiki/%EC%9A%A9%EC%96%B4-%EC%A0%95%EB%A6%AC)
* API는 용어정리에서 나온 단어를 활용하여 JSON 데이터를 만들도록 했습니다!
    * [API 정리 링크](https://github.com/Malloc72P/todo-list/tree/team-12-back/be)
* 설계된 API 중 일부분(컬럼 목록 조회하기, TodoLog 목록 조회하기)에 대한 MockUp API를 구현했습니다.
    * 엔티티를 수정하지 않으면서도 원하는 형식으로 클라이언트에게 JSON 데이터를 응답해줄 수 있도록, DTO를 적용해보았습니다!
* 스프링 앱서버와 프론트 웹서버를 AWS에 배포했습니다.
  * [프론트 주소](http://ec2-15-164-220-3.ap-northeast-2.compute.amazonaws.com:3000/)
  * [컬럼 목록보기 API](http://ec2-15-164-220-3.ap-northeast-2.compute.amazonaws.com:8080/api/columns)
  * 배포된 API에 대해선 포스트맨에 있는 팀 기능을 이용해서 프론트 분들과 공유할 수 있도록 했습니다!
![image](https://user-images.githubusercontent.com/41663269/114259476-21f4b300-9a09-11eb-9726-728c4f961465.png)


## 2. To do
* 주말동안 Spring Data JDBC를 학습하기
* 백엔드 서버에 Spring Data JDBC를 적용하고, API 기능 구현하기
* 순서가 보장되는 리스트를 DB에 저장하고 불러오는 방법 알아내서 적용하기(같은 컬럼 내에서 위 아래로 이동하는 기능 구현을 위해서!)

## 3. 궁금한 점
* 순서가 보장되는 리스트를 DB에 저장하고 불러오는 방법이 궁금합니다! 같은 컬럼에서 태스크를 위아래로 옮길려면 순서를 바꿔서 저장할 수 있어야 할 것 같은데, 이걸 DB에서 어떻게 해야 할 지 감이 안오네요 ㅠㅠ

Malloc72P committed:
feat : fe, be폴더를 생성

Malloc72P committed:
feat : 설명서에 제목을 추가하라

Malloc72P committed:
feat : 스프링 이니셜라이저를 이용해서 스프링 앱을 생성

Malloc72P committed:
feat : 엔티티 생성

- 목업 API 구현을 위해서 엔티티를 생성했다.
- 요구사항을 분석한 내용을 가지고 생성함

Malloc72P committed:
feat : 컨트롤러 클래스 생성

Malloc72P committed:
fix : 누락된 어노테이션 추가

Malloc72P committed:
docs : API 문서 초안 작성

- 컬럼 목록조회와 태스크 생성부분까지 작성함

Malloc72P committed:
docs : API문서를 작성

- 클라이언트에서 필요해보이는 대부분의 API에 대한 문서를 작성함

Malloc72P committed:
docs : 오타 수정

- 누락된 쌍따옴표를 추가함
- 누락된 유저필드를 추가함

Malloc72P committed:
docs : 예시로 작성한 패스워드를 수정

Malloc72P committed:
feat : 업데이트된 위키에 맞게 엔티티를 수정함

Malloc72P committed:
feat : show column list 목업 API를 추가

Malloc72P committed:
docs : 수정된 내용을 api문서에 반영

ksj55 committed:
docs : 엔티티 대소문자 수정

ksj55 committed:
feat : TodoLogController 구현하고 로그 조회하는 mockup API 추가

ksj55 committed:
feat : ColumDto와 TaskDto 추가해 응답포맷 최적화

ksj55 committed:
feat : TodoLogDto 추가해 응답포맷 최적화

ksj55 committed:
docs : API 문서 최신화

- 중복된 author data 보내는 문제를 해결

Malloc72P committed:
docs : 프로젝트에 대한 리드미 문서를 작성

Malloc72P committed:
Merge branch 'team-12-back' of https://github.com/Malloc72P/todo-list into team-12-back

Malloc72P committed:
chore : 안쓰는 import를 제거하고, 코드 스타일을 맞추기

Malloc72P committed:
chore : 안쓰는 컨트롤러의 메서드를 제거

ksundong commented on README.md:
```diff
@@ -1,2 +1,24 @@
-그룹프로젝트 #1
+
+
+# [Team-12] todo-list
+
+
+
+### 용어 정리
+
+* [용어 정리 링크](https://github.com/Malloc72P/todo-list/wiki/%EC%9A%A9%EC%96%B4-%EC%A0%95%EB%A6%AC)
+
+### 팀 규칙
+
+* [팀 규칙 링크](https://github.com/Malloc72P/todo-list/wiki/%ED%8C%80-%EA%B7%9C%EC%B9%99)
+
+### 백엔드 API 설계
+
+* [링크](https://github.com/Malloc72P/todo-list/tree/team-12-back/be)
+
+### 프론트 컴포넌트 설계
+
+* [링크](https://github.com/Malloc72P/todo-list/tree/team-12-front/fe)
+
```
README의 작성 아주 좋습니다.

기간은 촉박하지만, https://github.com/matiassingers/awesome-readme
이런 Repository를 참고하셔서 좀 더 좋은 README를 지속적으로 지향해주시는게 좋을 것 같아요.

Malloc72P replied:
우와.. 리드미가 엄청 세세하네요!! 저희도 이거 본받아서 작성하도록 하겠습니다. 😄

rla36 replied:
링크 감사합니다!! 꼭 읽어볼게요ㅎㅎ

ksundong commented on be/readme.md:
```diff
@@ -0,0 +1,230 @@
+# Todo 백엔드 서버
```
이 파일의 파일명은 소문잔데요, 대문자인 편이 더 좋을 것 같습니다.

Malloc72P replied:
넵 바로 반영하겠습니다!

ksundong commented on be/readme.md:
```diff
@@ -0,0 +1,230 @@
+# Todo 백엔드 서버
+
+* Todo 백엔드 서버입니당
+
+
+
```
빈 라인은 한 줄만 작성해주셔도 됩니다.

Malloc72P replied:
넵!

ksundong commented on be/readme.md:
```diff
@@ -0,0 +1,230 @@
+
+
+
+# API Document
```
API 문서화는 여러 도구를 이용할 수 있는데요.
- markdown, asciidoc 문서로 작성하는 경우(Post man API문서화도 이러한 방식입니다.)
- OAS, Swagger를 이용하는 경우
- Spring에서 제공하는 Spring REST Docs를 이용하는 경우
가 일반적이라고 생각됩니다.

각각 장단점이 있으니, 코드스쿼드 프로젝트를 수행하면서 한 번씩 경험해보시면 좋을 것 같습니다.

Malloc72P replied:
오! 여유가 생기면 이것도 한번 해보고 싶어요!!! 저번 QnA게시판에서는 스웨거를 썼던 것 같은데, 제대로 활용해보지 못해서 아쉬웠는데, 이번에는 활용해볼 수 있도록 해볼께요!

ksundong commented on be/readme.md:
```diff
@@ -0,0 +1,230 @@
+
+##### Request Body
+
+* userId
+  * ex : "모던 자바스크립트 예제 실습"
+* password
+  * ex : "1장 예제 내용 실습 후 커밋까지"
```
Table 형태로 만들어보는건 어떨까요?
타입이나, 제약사항, 예제 정도를 나열해주면 좋을 것 같습니다.
그리고 예시가 좀 부적절하다고 생각이 되네요!

Malloc72P replied:
정리가 깔끔한 것 같지 않아서 계속 아쉬웠는데, 테이블로 작성해보면 좋겠네요!

그리고 저 예시는 제가 실수했네요 ㅋㅋ 복붙하다가 수정하지 못했나봅니다! 바로 수정할께요!!

ksundong commented on be/readme.md:
```diff
@@ -0,0 +1,230 @@
+
+* /api/users/logout
+
+
```
여기는 따로 설명이 없네요 ^^;

Malloc72P replied:
추가하도록 하겠습니다 ㅎㅎ

ksundong commented on be/readme.md:
```diff
@@ -0,0 +1,230 @@
+
+* 사용자가 소유한 TodoList의 모든 컬럼을 응답합니다
+* 로그인된 사용자만 이용할 수 있는 API이다.
+* 따라서, URL이나 Request Body에 유저의 아이디를 넣을 필요가 없습니다.
```
설명 좋습니다.

ksundong commented on be/readme.md:
```diff
@@ -0,0 +1,230 @@
+* 로그인된 사용자만 이용할 수 있는 API이다.
+* 따라서, URL이나 Request Body에 유저의 아이디를 넣을 필요가 없습니다.
+
+##### JSON 포맷
```
Response여도 좋을 것 같아요.

ksundong replied:
그리고 꼭 json 포맷을 보여주지 않고 Request Field를 나열하던 방식으로 보여줘도 괜찮다고 생각합니다.
그에 대한 표기법도 있는데요. jsonPath에 대해서 한 번 알아보세요~

Malloc72P replied:
오 JSON포맷보다는 Response가 더 와닿는 것 같아요! 바로 반영해보고 싶어요!

그리고 jsonPath에 대해서는 [요기](https://www.lesstif.com/dbms/jsonpath-54952418.html)를 참조해서 공부해봤어요! 이걸 쓰면 프론트분들이랑 Response에 대해서 좀 더 편하게 커뮤니케이션할 수 있을 것 같다는 생각이 드네요! 알려주셔서 감사합니다!

ksundong replied:
네 온라인에서 jsonPath 분석도 가능하게 해주는 사이트도 있으니 적절히 사용하시면 좋을 듯하네요!

ksundong commented on be/todo/.gitignore:
```diff
@@ -0,0 +1,37 @@
+HELP.md
```
Spring Boot Starter가 생성해준 ignore파일인 것 같네요.
여기에 추가로 제외할 파일이 더 필요할 수도 있습니다.
이런 것들은 다 프리셋이 이미 있으니 gitignore.io에서 한 번 둘러보시는 것도 추천합니다.

Malloc72P replied:
`gitignore`는 생각하지 못했었는데 알려주셔서 감사해요! 필요한거 좀 더 있는 것 같은데 이것도 추가해야 겠네요.

ksundong commented on be/todo/build.gradle:
```diff
@@ -0,0 +1,22 @@
+}
+
+dependencies {
+	implementation 'org.springframework.boot:spring-boot-starter-web'
+	testImplementation 'org.springframework.boot:spring-boot-starter-test'
```
필요한 의존성은 미리 다 넣어줘도 됩니다.
spring data jdbc starter 의존성이 누락된 것 같아요.

Malloc72P replied:
앗 맞아요! 다음 주에 DB작업하면서 추가하려고 했어요! 😄

ksundong replied:
네 필요한 시점에 추가시켜주셔도 되지만, 연동이 생각보다는 쉽지 않은 경우가 많아서 유의하시면 좋을 것 같아요!

ksundong commented on be/todo/src/main/java/com/codesquad/todo/web/controller/ColumnController.java:
```diff
@@ -0,0 +1,35 @@
+public class ColumnController {
+
+    @GetMapping
+    public List<ColumnDto> showColumnList() {
```
보통 DTO자체가 약어여서 ColumnDTO로 표기하는 것이 더 정확하다고 생각됩니다.
하지만 이런건 그냥 팀에서 가져가면 되는 문제라서 큰 상관은 없어요~

ksundong replied:
반환타입이 `List<>`인데요, 가급적이면 key, value 형태로 반환해주시는 것이 변경에 더 유연하다는 점이 있습니다.
약간 억지를 부려서 List의 크기도 반환해주면 좋겠다라는 요구사항이 등장하게 되는경우, 객체형태라면 필드만 추가해주면 되지만, 이런 형태는 API의 Breaking Change가 발생하게됩니다.

API가 Breaking Change가 발생하게 되는 상황을 최대한 지양할 수 있는 것이 API를 설계하는데 있어서 중요하다고 생각됩니다.
한 번 생각해봐주세요!

ksundong replied:
여기서 Breaking Change는 이전과는 아예 다른 형태가 되는 것을 의미합니다.

Malloc72P replied:
아하 이거는 맵을 쓰면 되는거였군요! 처음에 저희도 
```
{ 
   columns : [ {...}, {...} ] 
}
``` 
이런식으로 하려다가 어떻게 해야할지 몰라서 API 설계를 변경했었어요! 근데 이거 맵을 쓰면 되는거였구나... 알려주셔서 감사합니다!
그리고 이렇게 하는 방법의 장점은 잘 모르고 있었는데, 이렇게 하면 요구사항의 변경에 유연하게 대처할 수 있겠군요! 정말 좋네요!

ksundong replied:
맵을 쓰기보다는 별도의 DTO 클래스를 정의해서 객체로 반환해주시면 됩니다!
저도 최근에 공부하면서 알게되었네요 ㅎㅎ

ksundong commented on be/todo/src/main/java/com/codesquad/todo/web/controller/ColumnController.java:
```diff
@@ -0,0 +1,35 @@
+        User author = new User(1, "라쿠운", "Racoon", "1234", "");
+
+        Column todoColumn = new Column(1, "TODO");
+        todoColumn.getTaskList().add(new Task(1, "task1", "taskContent1", author, todoColumn));
+        todoColumn.getTaskList().add(new Task(2, "task2", "taskContent2", author, todoColumn));
```
List를 꺼내서 넣기보다는, 넣을 객체를 Column에게 전달만 해줍시다!

Malloc72P replied:
아하 그러면 `todoColumn.addTask( )`같이 해주면 좋을 것 같아요! 
getter로 값을 가져와서 하기보다는, 컬럼 객체한테 요구하도록 요청해야 맞는 것 같네요!

ksundong commented on be/todo/src/main/java/com/codesquad/todo/web/controller/ColumnController.java:
```diff
@@ -0,0 +1,35 @@
+        inProgressColumn.getTaskList().add(new Task(3, "task3", "taskContent3", author, inProgressColumn));
+        columnList.add(new ColumnDto(inProgressColumn));
+
+        return columnList;
```
Mock API의 작성 훌륭합니다.

ksundong commented on be/todo/src/main/java/com/codesquad/todo/web/controller/TaskController.java:
```diff
@@ -0,0 +1,19 @@
+@RequestMapping("/api/tasks")
+public class TaskController {
+
+    @GetMapping()
```
ColumnController와는 일관되지않게 코드가 작성되었네요.
한 번 신경써주세요!

Malloc72P replied:
에고 실수했네요! 바로 수정할께요!

ksundong commented on be/todo/src/main/java/com/codesquad/todo/web/domain/Column.java:
```diff
@@ -0,0 +1,28 @@
+import java.util.List;
+
+public class Column {
+    private long id;
```
이정도는 미래지향적으로 설계해줘도 괜찮다고 생각합니다
`Long`형 어때요?

Malloc72P replied:
Long으로 해주는게 좋을 것 같아요! 테스트 프로젝트에서 이걸로 한번 당황한적이 있는데, 프리미티브 타입으로 하니까 커맨드 객체 매핑할때 id를 빼놓으니까 에러가 나더라구요! Null을 넣을 수 있는 Wrapper타입으로 하는게 좋을 것 같습니다!

ksundong commented on be/todo/src/main/java/com/codesquad/todo/web/domain/Column.java:
```diff
@@ -0,0 +1,28 @@
+    private String columnTitle;
+    private List<Task> taskList;
+
+    public Column(long id, String columnTitle) {
+        this.id = id;
+        this.columnTitle = columnTitle;
+        taskList = new ArrayList<>();
+    }
```
전체 인자를 다 받는 생성자를 만들어두고, 이를 호출하는 생성자의 형태로 바꿔볼까요?
생성자가 하나 더 추가되면 좋을 것 같아요.

Malloc72P replied:
앗 이 부분은 잘 이해하지 못했는데, 이렇게 말씀하신게 맞을까요?
```java
public Column(long id, String columnTitle, List<Task> taskList) {
    this.id = id;
    this.columnTitle = columnTitle;
    this.taskList = taskList;
}
public Column(long id, String columnTitle) {
    this(id, columnTitle, new ArrayList<>());
}
```
혹시 이렇게 하면 어떤 장점이 있을까요? 나중에 taskList를 직접 넣는 경우를 대비해서 이렇게 해두는건가요?

ksundong replied:
넵 정확합니다~ 생성자는 여러 방식으로 제공해주는 편이 유연한 설계를 도울 수 있다고 합니다!
제가 말씀드린게 꼭 정답은 아닐 수 있으니, 참고만 해주세요!

ksundong commented on be/todo/src/main/java/com/codesquad/todo/web/domain/Task.java:
```diff
@@ -0,0 +1,54 @@
+import java.time.LocalDateTime;
+
+public class Task {
+    private long id;
+    private String taskTitle;
+    private String taskContent;
+    private User author;
+    @JsonIgnore
+    private Column column;
+    private LocalDateTime createdDateTime;
+    private LocalDateTime updatedDateTime;
```
```suggestion
    private long id;
    private String taskTitle;
    private String taskContent;
    private User author;
    private LocalDateTime createdDateTime;
    private LocalDateTime updatedDateTime;
    
    @JsonIgnore
    private Column column;
```

ksundong replied:
User는 Ignore 되지 않아도 괜찮나요?
또한, toString이 들어갈 때에도 주의를 해주셔야할 것 같아요!

Malloc72P replied:
앗 맞습니다! 저희가 미처 신경쓰지 못한 것 같아요! 
디온이 말씀해주신대로, User도 Ignore되어야 합니다! 
해당 정보를 JSON으로 담을 필요가 없기 때문이예요! 
또 API설계에서도 빼고 전송하기도 하구요. 

저 근데 한가지 궁금한 점이 있어요! JSON으로 응답할때는 DTO를 쓰니까, 컬럼에 대한 JsonIgnore도 지워주려고 해요! 
그래도 괜찮은 설계인지 잘 모르겠어요! 
일단 저희는 엔티티에서는 JsonIgnore같은걸 다 빼주려고 합니다.
어떻게 하는게 좋을까요?

ksundong replied:
Ignore를 시켜주는 이유가 명확하네요!
사실 Domain에서는 굳이 JsonIgnore를 시켜줄 이유가 전혀 없죠!
DTO에서만 적용시켜주셔도 됩니다.
다만, toString을 적용하실 때는 여전히 주의하셔야한다는 점 염두에 두세요!

ksundong commented on be/todo/src/main/java/com/codesquad/todo/web/service/dto/ColumnDto.java:
```diff
@@ -0,0 +1,39 @@
+        this.taskList = convertTasksToTaskDtos(column.getTaskList());
+    }
+
+    private static List<TaskDto> convertTasksToTaskDtos(List<Task> taskList) {
+        List<TaskDto> taskDtoList = new ArrayList<>();
+        for (Task task : taskList) {
+            taskDtoList.add(new TaskDto(task));
+        }
+        return taskDtoList;
+    }
```
- `static` 키워드를 제거해볼까요?
- stream을 활용한 코드로 개선해볼까요?

Malloc72P replied:
오 static을 키워드를 제거하는 이유는 task객체한테 Dto로의 변환을 요청하는 방식으로 설계하기 위함인가요?
``` java
TaskDTO taskDTO = task.convertToDTO();
```
그럼 위와 같이 해보면 될 것 같아요!

아 그리고 스트림도  활용해 보도록 하겠습니다!

Malloc72P replied:
column객체에게 taskList를 TaskDtoList로 변환해주세요~ 하는 메서드를 만들었어요!
또 스트림도 적용했습니다!
``` java
public List<TaskDto> convertTaskListToTaskDtoList() {
    return taskList.stream().map(TaskDto::new).collect(Collectors.toList());
}
```
이제 ColumnDto에서는 아래와 같이 쓰면 됩니다!
``` java
public ColumnDto(Column column) {
    /*...*/
    this.taskList = column.convertTaskListToTaskDtoList();
}
```

ksundong replied:
좋습니다!

ksundong commented on be/todo/src/main/java/com/codesquad/todo/web/service/dto/TaskDto.java:
```diff
@@ -0,0 +1,41 @@
+        this.id = task.getId();
+        this.taskTitle = task.getTaskTitle();
+        this.taskContent = task.getTaskContent();
+        this.authorName = task.getAuthor().getName();
```
디미터 법칙에 대해서 한 번 공부해보세요!
그리고, 이렇게 get을 chaining하게되면 NPE가 굉장히 발생하기 쉬워집니다.

https://woowacourse.github.io/javable/post/2020-06-02-law-of-demeter/
https://johngrib.github.io/wiki/law-of-demeter/

Malloc72P replied:
아하 그러면 이런 경우에는 `task.getAuthorName()`이라는 메서드를 task에 만들어주면 좋을 것 같아요! 그리고 아무 생각없이 get을 체이닝하는것도 피해서 구현해야겠네요!

Malloc72P replied:
아니면 task객체의 내부구조에 너무 깊게 결합되는걸 막기 위해서 `getAuthorName()`을 하지 않아도 되도록 설계를 바꿔주는 것도 좋을 것 같아요! 좀 더 고민해보고 적용해보겠습니다!

ksundong replied:
네 여러가지 잘 생각해보고 고민해보고 하면서 실력이 느실거에요! 👍

ksundong commented on be/todo/src/main/resources/application.properties:
```diff
@@ -0,0 +1 @@
+
```
아직은 별달리 설정할 게 없었나보네요!

Malloc72P replied:
네 아직 Spring Data Jdbc를 추가하지 않아서 아직까지는 설정할 내용이 없더라구요!
주말동안 Spring Data Jdbc 프로젝트를 만들어서 공부해보았는데, 다음 PR날릴땐 설정이 많이 들어갈 것 같아요 😆

ksundong commented on be/todo/src/test/java/com/codesquad/todo/TodoApplicationTests.java:
```diff
@@ -0,0 +1,13 @@
+class TodoApplicationTests {
+
+    @Test
+    void contextLoads() {
```
한 번 Domain 정도는 테스트 해보는 것도 좋은 것 같아요~
여유가 된다면 도전해보시죠!

Malloc72P replied:
꼭 해보겠습니다!

ksundong replied:
꼭 하실 필요는 없어요 ㅎㅎ

ksundong reviewed (approved):
안녕하세요! NAS & Jung!! 리뷰어 Dion입니다.
첫 프로젝트라 이런 저런 협업관련 시행착오가 있었을 것 같아요. 고생많으셨습니다.

- 용어 정리를 해두는건 매우 중요하죠. 공통의 도메인 용어를 사용해야 의사소통을 하는데 문제가 없게됩니다. 잘하셨어요. 👍 
- API 문서화도 좋은 시도입니다. 다만, 더 좋은 목표를 향해서 고민해봅시다.
- Mock API를 제공해주는것도 매우 바람직합니다.
- DTO의 적용도 좋습니다. 아직은 Service Layer가 분리되지 않았지만, Controller에 Domain이 노출되지 않도록 고민해봐주세요.
  - 특히 생각하신 부분이 매우 정확합니다. 엔티티는 구현 요구사항에 영향을 받으면 안됩니다.
  - 거기에 더해서 API의 breaking change도 발생하지 않는 구조를 고민해봐주세요.
- API 문서화를 Postman을 이용해서 해주셨군요. 잘 하셨습니다. Postman은 정말 강력한 도구인데요. 잘 활용해서 사용해보셔도 좋을 것 같아요.
- Spring Data JDBC는 저희가 쓸 때에는 정말 불편한 부분이 있었는데, 지금은 많이 발전해서 사용성이 좋아졌다고 생각됩니다. 잘 활용해보세요!
- 사실 궁금한 점이 설계의 핵심이라서 조언을 드리기가 참 애매한 것 같아요. 다음주에도 고민해보시고 정말 안풀리면 코멘트나 DM 주세요.

고생하셨습니다! 다음주에는 더 많은 공부를 하셨으면 좋겠네요!
코멘트 참고해주시고, 변경요청 드릴 부분은 많지 않아서 approve 하겠습니다!

Malloc72P committed:
docs : 백엔드 리드미 문서의 양식 수정

- 빈 라인은 한줄만 띄우도록 수정
- 대문자로 파일 이름 수정하려고 시도

Malloc72P committed:
docs : readme.md파일의 이름을 README.md로 수정

Malloc72P committed:
docs : API문서를 보강

- 로그인, 로그아웃에 대한 설명과 리스폰스를 추가
- JSON 포맷 대신 리스폰스로 고쳐씀
- 리스트 대신 테이블로 요청메서드와 URL, 그리고 Request Body를 표기함.

Malloc72P committed:
docs : 누락된 수정내용을 추가

Malloc72P committed:
chore : gitignore 업데이트

- windows와 java 관련 태그로 ignore할 내용을 추가

Malloc72P committed:
reformat : 프리미티브 대신 Wrapper 타입의 Long으로 id의 타입 수정

Malloc72P committed:
chore : 불필요한 괄호 제거

Malloc72P committed:
fix : Long타입이므로 숫자뒤에 L을 붙여주도록 수정

Malloc72P committed:
feat : 컬럼객체의 메서드를 통해 태스크를 추가할 수 있도록 구현

Malloc72P committed:
reformat : taskList를 dtoList로 변환하는 메서드를 멤버 메서드로 수정

- static보다는 객체한테 요청해서 결과를 획득할 수 있도록 구현을 수정함
- 추가로 stream을 사용해서 코드를 깔끔하게 고침

Malloc72P committed:
feat : 컨트롤러에서 List대신 Map으로 응답하도록 수정함

- 리스트로 응답하면 Breaking Change될 확률이 커진다.
- 추가데이터를 끼워넣을 방법이 없기 때문이다.
- 맵으로 하면 이 문제가 해결된다. 그냥 키 하나 더 추가하고 넣으면 되니까 그렇다.
- 여하튼 이런 이유로 맵으로 응답하도록 수정했다.

Malloc72P committed:
docs : api문서 최신화

- 리스트에서 맵형식으로 응답이 바뀌어서 문서를 최신화함

Malloc72P committed:
docs : 누락된 쌍따옴표 추가함

Malloc72P committed:
feat : 전체 인자를 다 받는 생성자를 만들어보자

Malloc72P committed:
reformat : intelliJ의 리포맷 기능을 수행

Malloc72P commented:
안녕하세요 디온! 코멘트 달아주신 내용을 다 반영했어요!

한가지 궁금한 점이 남았는데요, https://github.com/codesquad-members-2021/todo-list/pull/14#discussion_r611136413 이 부분이예요!
전체 인자를 다 받는 생성자를 만들어야 하는 이유을 잘 모르겠어요!
전체 인자를 다 받은 생성자를 만드는 이유는 나중에 빌더패턴을 적용하기 쉽게 하기 위한걸까요? 
요 부분에 대한 힌트 하나만 주세요! 😄

ksundong commented:
큰 이유는 없고, 위에 말씀드린바와 같이, 유연한 사용을 위함이라고 말씀드리고 싶습니다!
반드시 빈 컬럼만 생성되어야 할 이유가 없지 않을까요?

ksundong commented on be/README.md:
```diff
@@ -21,6 +21,17 @@
+
+* 로그인을 요청하는 API입니다
+* 해당 API를 통해 로그인을 해야 `컬럼목록조회`등의 로그인이 필요한 기능을 이용할 수 있습니다.
+* **아직 미구현된 API입니다만, 이런 형식으로 구현할 예정입니다!**
```
요것도 어떻게하면 더 강조가 될 지 생각해봅시다.
저는 제목에다가 넣는 편이 더 강조가 되는 것 같아서 (미구현)이라고 대문짝만하게 적어뒀었네요 ㅋㅋ

ksundong commented on be/todo/src/main/java/com/codesquad/todo/web/domain/Column.java:
```diff
@@ -25,4 +25,8 @@ public String getColumnTitle() {
     }
+
+    public void addTask(Task task) {
+        taskList.add(task);
```
자 그러면 하나 더.. task에도 바로 이 column을 지정해주는 방식은 어떨까요?
그리고, task에서 이러한 내용을 관리하는 것이 좋을지, column에서 관리하는 것이 좋을지 생각해봅시다.

ksundong commented on be/todo/src/main/java/com/codesquad/todo/web/domain/Column.java:
```diff
@@ -29,4 +32,8 @@ public String getColumnTitle() {
     }
+
+    public List<TaskDto> convertTaskListToTaskDtoList() {
+        return taskList.stream().map(TaskDto::new).collect(Collectors.toList());
```
stream 이후로 한 줄 씩 개행해주시면 더 가독성이 높아집니다!

ksundong commented on be/todo/src/main/java/com/codesquad/todo/web/controller/ColumnController.java:
```diff
@@ -9,27 +9,33 @@
     @GetMapping
-    public List<ColumnDto> showColumnList() {
-        List<ColumnDto> columnList = new ArrayList<>();
+    public Map<String, Object> showColumnList() {
```
Map을 반환하기 보다는 별도의 DTO 클래스를 정의해서 넘겨줍시다!

ksundong commented on be/README.md:
````diff
@@ -81,44 +81,49 @@
   ```
   
 
+* `21-04-11 오후 05:19`부로 columns키가 추가되었습니다
+  * [수정 사유는 링크와 같습니다](https://github.com/Malloc72P/todo-list/issues/24)
+
````
breaking change에 대한 전달까지.. 💯

ksundong reviewed (approved):
리뷰내용 반영 빠르게 해주셨네요!
고생하셨습니다!
